[YouTube] Add support for poTokens

Stypox opened the pull request:
#### What is it?
- [x] Bugfix (user facing)

#### Description of the changes in your PR

General information about poTokens and about this PR structure:
- YouTube now requires integrity checks to access their clients. The most "vulnerable" client is the WEB client, since they can't enforce integrity checks on all web browsers, so that's the only client (for now) that we have found a way to obtain an integrity token for.
- In order to obtain a poToken, we need to run BotGuard, an obfuscated virtual machine implemented in JavaScript that performs the integrity checks and gives us an integrity token. In order to make the integrity checks succeed, we need to run this VM in an environment that resembles a browser as much as possible. The integrity token can be used to generate multiple poTokens. Two network requests are needed: `Create` to obtain the VM code, `GenerateIT` to obtain the integrity token after running the VM code. See the README [here](https://github.com/LuanRT/BgUtils) for the detailed steps.
- `PoTokenGenerator` is the base class for all poToken generators. It has a factory method that allows asynchronously obtaining a new instance of a `PoTokenGenerator`, and then two methods to generate a poToken given a specific identifier, and a method to check if the integrity token has expired.
- `PoTokenWebView` is currently the only implementation of `PoTokenGenerator`, but we might want to add other implementations in the future, e.g. ones that do not rely on WebView.
- `PoTokenProviderImpl` implements the extractor interface and is supposed to take care of possibly multiple `PoTokenGenerator`s (although right now there is only one based on WebView). It takes care of retrying in case of problems, recreates a new `PoTokenGenerator` if the current one expired, and finally returns a `PoTokenResult`. A `PoTokenResult` contains two poTokens: one for the specific requested video id (used to fetch the player), and another that can be generated only once as the first thing and is specific to a visitor data (used in streaming urls).

TODO:
- [x] The JavaScript poToken implementation comes from https://github.com/LuanRT/BgUtils 
- [x] Obtaining a poToken via WebView
- [x] ~Obtaining a poToken with something like [HtmlUnit](https://www.htmlunit.org/gettingStarted.html)~ not doable unfortunately
- [x] Handling devices that don't have a WebView (needs to be tested)
- [x] Passing the poToken to the extractor when requested
- [x] Passing the poToken to player network requests (not sure if needed?)
- [ ] Change challenge fetching to InnerTube and sending integrity result to `youtube.com` like the desktop website does, this avoids requirement of the `jnn-pa.googleapis.com` domain - partially done, InnerTube attestation endpoint is not used yet
- [ ] Understand whether we need to change user agent everywhere

You can test whether the poTokens generated work also using the latest yt-dlp commit from their git repo (older commits won't work!), this way (take PLAYER_POT, STREAMING_POT and VISITOR_DATA from logcat):
```bash
yt-dlp "https://www.youtube.com/watch?v=i_SsnRdgitA" --extractor-args 'youtube:player_client=web;player-skip=webpage,configs;po_token=web.player+PLAYER_POT,web.gvs+STREAMING_POT;visitor_data=VISITOR_DATA'
```

#### Fixes the following issue(s)
- Fixes #11803

#### Relies on the following changes
- TeamNewPipe/NewPipeExtractor#1272

#### APK testing
The APK can be found by going to the "Checks" tab below the title. On the left pane, click on "CI", scroll down to "artifacts" and click "app" to download the zip file which contains the debug APK of this PR. You can find more info and a video demonstration [on this wiki page](https://github.com/TeamNewPipe/NewPipe/wiki/Download-APK-for-PR).


#### Due diligence
- [x] I read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md).

Stypox commented:
Now the PR builds fine based on https://github.com/TeamNewPipe/NewPipeExtractor/pull/1247, you can download the APK which uses poTokens! Let us know if you notice any issues.

Stypox commented on app/src/main/java/org/schabi/newpipe/util/potoken/PoTokenWebView.kt:
```diff
@@ -0,0 +1,309 @@
+        private val TAG = PoTokenWebView::class.simpleName
+        private const val GOOGLE_API_KEY = "AIzaSyDyT5W0Jh49F30Pqqtyfdf7pDLFKLJoAnw"
+        private const val REQUEST_KEY = "O43z0dpjhgX20SCx4KAo"
+        private const val USER_AGENT = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.3"
```
No, for some reason it does not work with the Firefox user agent. It would work with the curl user agent though, I don't know why...

ale5000-git replied:
Perhaps checks on Firefox are stricter.
Just a deduction of what may be the cause:
Some webservers may also check the "case" of headers (User-Agent vs user-agent; this also apply to other headers) and even the order in which the headers are sent to the server may matter.
The WebView is more likely to have an implementation similar to Chrome, so it will be more likely to fail with FireFox User-Agent since the implementation differs completely.

Profpatsch commented on app/src/main/assets/po_token.html:
```diff
@@ -0,0 +1,211 @@
+      throw new Error(`[BotGuardClient]: Failed to load program (${error.message})`);
+    }
+
+    // an asynchronous function runs in the background and it will eventually call
+    // `vmFunctionsCallback`, however we need to manually tell JavaScript to pass
+    // control to the things running in the background by interrupting this async
+    // function in any way, e.g. with a delay of 1ms. The loop is most probably not
+    // needed but is there just because.
+    for (let i = 0; i < 10000 && !this.vmFunctions.asyncSnapshotFunction; ++i) {
+      await new Promise(f => setTimeout(f, 1))
+    }
```
I … don’t think this is how async works. The timeout is just gonna be scheduled on a new task, but the code before the loop still runs on a microtask on the previous task.

Stypox replied:
The loop actually executes only once as far as I know, I still put a loop because you never know

Profpatsch commented:
Can there be an architecture overview of this somewhere? From a skim of the code I don’t get any idea of what problem this solves or how the solution is structured.

Stypox commented:
- YouTube now requires integrity checks to access their clients. The most "vulnerable" client is the WEB client, since they can't enforce integrity checks on all web browsers, so that's the only client (for now) that we have found a way to obtain an integrity token for.
- In order to obtain a poToken, we need to run BotGuard, an obfuscated virtual machine implemented in JavaScript that performs the integrity checks and gives us an integrity token. In order to make the integrity checks succeed, we need to run this VM in an environment that resembles a browser as much as possible. The integrity token can be used to generate multiple poTokens. Two network requests are needed: `Create` to obtain the VM code, `GenerateIT` to obtain the integrity token after running the VM code. See the README [here](https://github.com/LuanRT/BgUtils) for the detailed steps.
- `PoTokenGenerator` is the base class for all poToken generators. It has a factory method that allows asynchronously obtaining a new instance of a `PoTokenGenerator`, and then two methods to generate a poToken given a specific identifier, and a method to check if the integrity token has expired.
- `PoTokenWebView` is currently the only implementation of `PoTokenGenerator`, but we might want to add other implementations in the future, e.g. ones that do not rely on WebView.
- `PoTokenProviderImpl` implements the extractor interface and is supposed to take care of possibly multiple `PoTokenGenerator`s (although right now there is only one based on WebView). It takes care of retrying in case of problems, recreates a new `PoTokenGenerator` if the current one expired, and finally returns a `PoTokenResult`. A `PoTokenResult` contains two poTokens: one for the specific requested video id (used to fetch the player), and another that can be generated only once as the first thing and is specific to a visitor data (used in streaming urls).

Let me know which places are not documented enough.

Profpatsch commented:
@Stypox I think it would be good to include this documentation into the source code somewhere, maybe in the interface module.

Profpatsch commented:
So that people who want to understand the code later don’t have to find this PR and looks through lots of issues first

Profpatsch commented:
As a general comment, I don’t think it’s wise to add interfaces before even having need of a second implementation, it just makes the code more indirect and harder to read than necessary.

Profpatsch reviewed (approved):
I get an “Couldn get HLS manifest” Extraction Exception, but the stream extraction seems to work just fine, so LGTM I’d say so we can create a release.

Profpatsch commented on app/build.gradle:
```diff
@@ -208,7 +208,7 @@ dependencies {
     // WORKAROUND: if you get errors with the NewPipeExtractor dependency, replace `v0.24.3` with
     // the corresponding commit hash, since JitPack is sometimes buggy
-    implementation 'com.github.TeamNewPipe:NewPipeExtractor:v0.24.4'
+    implementation 'com.github.FireMasterK:NewPipeExtractor:d2cbd09089e8af933738f98b671ad58236a79d6e'
```
Oh I missed that the NewPipeExtractor still points to an unofficial repo, that needs to be fixed before merge

Stypox commented:
> As a general comment, I don’t think it’s wise to add interfaces before even having need of a second implementation, it just makes the code more indirect and harder to read than necessary.

Yes I totally agree with you, but when I started working on this I hoped I would be able to create two implementations

Isira-Seneviratne commented on app/src/main/java/org/schabi/newpipe/util/potoken/PoTokenWebView.kt:
```diff
@@ -0,0 +1,356 @@
+) : PoTokenGenerator {
+    private val webView = WebView(context)
+    private val disposables = CompositeDisposable() // used only during initialization
+    private val poTokenEmitters = mutableListOf<Pair<String, SingleEmitter<String>>>()
```
```suggestion
    private val poTokenEmitters = ConcurrentHashMap<String, SingleEmitter<String>>()
```

Stypox replied:
No, a list here is totally fine. Small lists are much faster that hash maps and I don't expect this list to grow bigger than a couple of items.

Isira-Seneviratne commented on app/src/main/java/org/schabi/newpipe/util/potoken/PoTokenWebView.kt:
```diff
@@ -0,0 +1,356 @@
+     * the right emitters.
+     */
+    private fun addPoTokenEmitter(identifier: String, emitter: SingleEmitter<String>) {
+        synchronized(poTokenEmitters) {
+            poTokenEmitters.add(Pair(identifier, emitter))
+        }
```
```suggestion
        poTokenEmitters[identifier] = emitter
```

Isira-Seneviratne commented on app/src/main/java/org/schabi/newpipe/util/potoken/PoTokenWebView.kt:
```diff
@@ -0,0 +1,356 @@
+     * or an error.
+     */
+    private fun popPoTokenEmitter(identifier: String): SingleEmitter<String>? {
+        return synchronized(poTokenEmitters) {
+            poTokenEmitters.indexOfFirst { it.first == identifier }.takeIf { it >= 0 }?.let {
+                poTokenEmitters.removeAt(it).second
+            }
+        }
```
```suggestion
        return poTokenEmitters.remove(identifier)
```

Isira-Seneviratne commented on app/src/main/java/org/schabi/newpipe/util/potoken/PoTokenProviderImpl.kt:
```diff
@@ -0,0 +1,117 @@
+
+        val (poTokenGenerator, visitorData, streamingPot, hasBeenRecreated) =
+            synchronized(WebPoTokenGenLock) {
+                val shouldRecreate = webPoTokenGenerator == null || forceRecreate ||
+                    webPoTokenGenerator!!.isExpired()
```
```suggestion
                val shouldRecreate = forceRecreate || webPoTokenGenerator?.isExpired() == true
```

Stypox replied:
I know this is a possible syntax, but I preferred making the null check clear to have better clarity

Isira-Seneviratne commented on app/src/main/java/org/schabi/newpipe/util/potoken/PoTokenWebView.kt:
```diff
@@ -0,0 +1,356 @@
+         */
+        private fun runOnMainThread(
+            emitterIfPostFails: SingleEmitter<out Any>,
+            runnable: () -> Unit,
```
```suggestion
            runnable: Runnable,
```

Isira-Seneviratne commented on app/src/main/java/org/schabi/newpipe/util/potoken/PoTokenGenerator.kt:
```diff
@@ -0,0 +1,35 @@
+     * `webPoSignalOutput` previously obtained in the initialization of [PoTokenWebView]. Can be
+     * called multiple times.
+     */
+    fun generatePoToken(identifier: String): Single<String>
```
Couldn't this be written as a suspend function instead? A `runBlocking` version could also be added to call this from Java.

Stypox replied:
I guess we can change it to coroutines when we convert this to the refactor branch, but I don't think we have coroutines anywhere in NewPipe non-refactor, so I didn't want to introduce new libraries and new ways of writing code, so I just stuck to RxJava.

YT-Advanced commented on app/src/main/java/org/schabi/newpipe/util/potoken/PoTokenWebView.kt:
[SimpMusic](https://github.com/maxrave-dev/SimpMusic/commit/2a694e34834757d8f361b8159fc0e5eb4d47444a#diff-ab57e32311b93d90a5d5402186748de6ac592d11efc8b90b6e7512f364e39aa4) used a simper approach to obtain the potoken without webview

YT-Advanced replied:
> There are no playback errors (http 403, bot etc) now, this method can be explored later🤔. Maybe their application has not fixed the http 403 error. Maybe it's because of a bot error on the login.

[MediaServiceCore](https://github.com/yuliskov/MediaServiceCore/blob/master/youtubeapi/src/main/java/com/liskovsoft/youtubeapi/app/potoken/PoTokenApi.kt) also the the similar way (with retrofit instead of ktor)

YT-Advanced replied:
> Is it possible to add this without breaking current support?

Uhm jusst use it as a fallback method in case devices don't have webview itself

LuanRT replied:
WebPO tokens cannot be generated without a proper DOM and JavaScript engine. From a quick analysis of the code shared by @YT-Advanced, they're simply retrieving an obfuscated challenge and passing it to the Generate Integrity Token endpoint. That should yield a web-safe fallback token, but *it is not* a valid PoToken, and it is not bound to any session. `GenerateIT` expects the snapshot result from BotGuard — not the raw scrambled challenge.

"Another thing: do not hardcode PO Tokens. Their validity is determined by the integrity token's TTL

Stypox replied:
> Is it possible to add this without breaking current support?

Yes, I engineered the code to allow for multiple PoToken providers

> [MediaServiceCore](https://github.com/yuliskov/MediaServiceCore/blob/master/youtubeapi/src/main/java/com/liskovsoft/youtubeapi/app/potoken/PoTokenApi.kt) also the the similar way (with retrofit instead of ktor)

They seem to be using a [NodeJS V8 runtime](https://github.com/yuliskov/MediaServiceCore/blob/5125aa73e97a94de1654148ad3ea2cb649d2fdc9/youtubeapi/src/androidTest/java/com/liskovsoft/youtubeapi/app/potoken/PoToken.kt#L105), which would probably be really heavy, but we might consider it in the future. For now I tried with HtmlUnit but it wasn't feature-complete enough to run the JS code.

> [SimpMusic](https://github.com/maxrave-dev/SimpMusic/commit/2a694e34834757d8f361b8159fc0e5eb4d47444a#diff-ab57e32311b93d90a5d5402186748de6ac592d11efc8b90b6e7512f364e39aa4) used a simpler approach to obtain the potoken without webview

Interesting! The actual code is [here](https://github.com/maxrave-dev/SimpMusic/blob/28e6f09546ea8636bee3e7e19b88ff15dcc0493b/kotlinYtmusicScraper/src/main/java/com/maxrave/kotlinytmusicscraper/YouTube.kt#L1241). I do not understand where they run the JS code though (?), it seems like they just get the challenge from Create and pass it to GenerateIT directly, which I don't believe can work (?).

Stypox commented:
@spvkgn thanks for reporting, please try the latest APK, I should have fixed that. Your WebView seems to use a really old version of JavaScript that doesn't have the `globalThis` object, but in our case `this` should return the same global scope anyway when called inside of a global function.

Isira-Seneviratne commented on app/src/main/java/org/schabi/newpipe/util/potoken/PoTokenWebView.kt:
```diff
@@ -0,0 +1,376 @@
+        val webviewSettings = webView.settings
+        //noinspection SetJavaScriptEnabled we want to use JavaScript!
+        webviewSettings.javaScriptEnabled = true
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+            webviewSettings.safeBrowsingEnabled = false
+        }
```
The [compat version](https://developer.android.com/reference/androidx/webkit/WebSettingsCompat#setSafeBrowsingEnabled(android.webkit.WebSettings,boolean)) of this method might be needed to support older versions of Android.

More information: https://www.youtube.com/watch?v=7Wq8Lwt_Cbg

Stypox commented:
I reworked the JavaScript code to not use any new JS features to achieve compatibility with as many WebView implementations as possible (including the API 21 emulator I was testing this on, which previously did not work). To do this, I moved all conversion functions to Kotlin (since those do not need to run in JS, so it's better if they are written in Kotlin where errors are caught at compile time and/or with nice stacktraces) and used `Promise.then()` instead of async functions.

While at it, I added more ways to detect if the WebView is not able to actually run the code, by making it so that the poToken generator returns `null` if there are any uncaught (syntax) errors. This makes the extractor aware of the fact that no poTokens can be generated, and tries other methods. Note that uncaught errors are thrown only while parsing JS, because in all places where JS is actually executed there is a `try{}catch(){}` around it.

I tested on API 21 emulated, API 34 emulated, and Android 14 on-device, and it worked everywhere (with playable videos).

**Please test this ASAP, and report any issues. I will make a release tomorrow morning.**

> The [compat version](https://developer.android.com/reference/androidx/webkit/WebSettingsCompat#setSafeBrowsingEnabled(android.webkit.WebSettings,boolean)) of this method might be needed to support older versions of Android.

Thank you, solved.

PepeTux commented:
Tested on a 
Moto G8 Power, Android 11: ✅
Moto G82, Android 13: ✅
TV Box Rockchip, Android 11: ✅

Great, thanks 👍

tzagim commented:
works great on A15!
TNX

Stypox commented:
I am merging this now since we need to do a new release. This PR probably needs some refinement, and will especially need to adapt to future extractor changes (because the current extractor API around poTokens is going to be changed), but works well enough. Let's hope all poToken implementations behave well!

bodtx commented:
Working great on redmi note 8 pro 🙏

gbansaghi commented:
Tried the APK on a OnePlus 5 with LineageOS 22.1 (Android 15) and still got the "Sign in to confirm..." error.

bodtx commented:
> Tried the APK on a OnePlus 5 with LineageOS 22.1 (Android 15) and still got the "Sign in to confirm..." error.

Be aware that it is installing another instance of new pipe called" new pipe po token" 
Did you test the correct one, not the old one?

Stypox commented:
@gbansaghi this PR does not solve https://github.com/TeamNewPipe/NewPipe/issues/11139 (and probably there is no way to solve it). Also this PR is already shipped in release 0.27.6.

gbansaghi commented:
> Be aware that it is installing another instance of new pipe called" new pipe po token" Did you test the correct one, not the old one?

Yes, I tested with the new instance, same error.


> @gbansaghi this PR does not solve #11139 (and probably there is no way to solve it).


I reported my results because AudricV [commented on #11139](https://github.com/TeamNewPipe/NewPipe/issues/11139#issuecomment-2628993883):


> […] it would be great to let us know if the latest debug APK of #11955 changes anything for you.


And, well, it didn't, but I figured a negative data point is still a data point. (I did overlook the fact that it had been merged though)

FineFindus commented on app/src/main/java/org/schabi/newpipe/util/potoken/JavaScriptUtil.kt:
```diff
@@ -0,0 +1,113 @@
+        val descrambled = descramble(scrambled.getString(1))
+        JsonParser.array().from(descrambled)
+    } else {
+        scrambled.getArray(1)
```
I think this should be accessing index `0` and not `1`? At least it did so in the [previous JS version](https://github.com/Stypox/NewPipe/blob/2b183a057651a79aa826a110e3eebf126befdd6c/app/src/main/assets/po_token.html#L100).

Mafiadoener36 commented:
As the artifact expired - can someone PLZ create a ? New one?
